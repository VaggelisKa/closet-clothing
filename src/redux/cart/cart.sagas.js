import { takeLatest, call, put, all, select } from 'redux-saga/effects';
import userActionTypes from '../user/user.types';
import CartActionTypes from './cart.types';
import { getUserCartRef } from '../../firebase/firebase.utils';
import { selectCurrentUser } from '../user/user.selectors';
import { selectCartItems } from './cart.selectors';

import { clearCart, setCartFromFirebase } from './cart.actions';

export function* clearCartOnSignOut() {
  yield put(clearCart());
}

export function* updateCartInFirebase() {
  const currentUser = yield select(selectCurrentUser);
  if (currentUser) {
    try {
      const cartRef = yield getUserCartRef(currentUser.id);
      const cartItems = yield select(selectCartItems);

      yield cartRef.update({ cartItems });
    } catch (error) {
      // console.log(error);
    }
  }
}

export function* checkCartFromFirebase({ payload: user }) {
  const cartRef = yield getUserCartRef(user.id);
  const cartSnapshot = yield cartRef.get();
  yield put(setCartFromFirebase(cartSnapshot.data().cartItems));
}

export function* onSignInSuccess() {
  yield takeLatest(
    userActionTypes.SIGN_IN_SUCCESS,
    checkCartFromFirebase
  );
}

export function* onSignOutSuccess() {
  yield takeLatest(
    userActionTypes.SIGN_OUT_SUCCESS,
    clearCartOnSignOut
  );
}

export function* onCartChange() {
  yield takeLatest(
    [
      CartActionTypes.ADD_ITEM,
      CartActionTypes.REMOVE_ITEM_FROM_CART,
      CartActionTypes.CLEAR_CART
    ],
    updateCartInFirebase
  );
}

export function* cartSagas() {
  yield all(
    [
      call(onSignOutSuccess),
      call(onCartChange),
      call(onSignInSuccess)
    ]
  );
}
