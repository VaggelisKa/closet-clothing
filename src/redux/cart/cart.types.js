const CartActionTypes = {
    TOGGLE_CART_HIDDEN: 'TOGGLE_CART_HIDDEN',
    ADD_ITEM: 'ADD_ITEM',
    DECREASE_QUANTITY: 'DECREASE_QUANTITY',
    REMOVE_ITEM_FROM_CART: 'REMOVE_ITEM_FROM_CART',
    CLEAR_CART: 'CLEAR_CART'
}

export default CartActionTypes;