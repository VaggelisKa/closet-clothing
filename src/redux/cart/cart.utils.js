export const addItemToCart = (cartItems, cartItemToAdd) => {
    const doesExistInCart = cartItems.find(cartItem => cartItem.id === cartItemToAdd.id);

    if (doesExistInCart) {
        return cartItems.map(cartItem => 
                cartItem.id === cartItemToAdd.id
                ? {...cartItem, quantity: cartItem.quantity + 1} : cartItem 
            )
    }

    return [...cartItems, {...cartItemToAdd, quantity: 1}];
}

export const decreaseQuantity = (cartItems, cartItemToRemove) => {
    const existingCartItem = cartItems.find(item => item.id === cartItemToRemove.id);

    if (existingCartItem.quantity === 1) {
        return cartItems.filter(item => item.id !== existingCartItem.id);
    }

    return cartItems.map(item =>
            item.id === cartItemToRemove.id 
            ? {...item, quantity: item.quantity - 1} : item
        );
}