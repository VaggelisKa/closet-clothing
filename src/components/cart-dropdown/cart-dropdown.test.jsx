/* eslint-disable no-undef */
import React from 'react';
import { shallow, mount } from 'enzyme';

import CartDropdown from './cart-dropdown.component';
import Button from '../button/button.component';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

describe('CartDropdown component', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn()
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch
    };

    wrapper = shallow(<CartDropdown.WrappedComponent {...mockProps} />);
  });

  it('Should render cart dropdown component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should call history.push when button is clicked', () => {
    wrapper.find(Button).simulate('click');

    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });
});
