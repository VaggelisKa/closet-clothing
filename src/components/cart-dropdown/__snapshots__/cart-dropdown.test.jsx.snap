// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component Should render cart dropdown component 1`] = `
<div
  className="cart-dropdown"
>
  <div
    className="cart-items"
  >
    <Memo(CartItem)
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
    />
  </div>
  <Button
    id="button"
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </Button>
</div>
`;
